### ISSUE TITLE
fwupdmgr security level: detect zram

### ISSUE THREAD

**Issue by edwintorok** [fwupd/fwupd#2537]
**Describe the bug**
fwupdmgr security --force shows 'Linux swap: unencrypted' as Red, which is wrong.
Fedora 33 uses [zram as swap](https://fedoraproject.org/wiki/Changes/SwapOnZRAM), hence encrypting swap in this case would be pointless as its contents doesn't leave RAM.

**Steps to Reproduce**
On Fedora 33 run:
```
$ sudo fwupdmgr security --force
Host Security ID: HSI:2+! (v1.5.0)

HSI-1
✔ TPM v2.0:                      Found

HSI-2
✔ IOMMU:                         Enabled
✔ TPM PCR0 reconstruction:       Valid

HSI-3
✘ Suspend-to-idle:               Disabled
✘ Suspend-to-ram:                Enabled

Runtime Suffix -U
✘ Firmware updates:              Not supported

Runtime Suffix -A
✘ Firmware attestation:          Not supported

Runtime Suffix -!
✔ Linux kernel:                  Untainted
✔ Linux kernel lockdown:         Enabled
✔ fwupd plugins:                 Untainted
✘ Linux swap:                    Unencrypted

This system has HSI runtime issues.
 » https://github.com/fwupd/fwupd/wiki/Host-security-ID-runtime-issues

$ sudo swapon -s
Filename				Type		Size	Used	Priority
/dev/zram0                             	partition	4194300	0	100
```

**Expected behavior**
If all swap devices are zram (or if there are no swap devices) then this should be treated as green,
and perhaps with a different text than unecrypted, e.g. say "Compressed RAM only"

**fwupd version information**
```shell
$ fwupdmgr --version
client version:	1.5.0
compile-time dependency versions
	gusb:	0.3.5

daemon version:	1.5.0
```

Please note how you installed it (`apt`, `dnf`, `pacman`, source, etc):
dnf install

**fwupd device information**
Please provide the output of the fwupd devices recognized in your system.

```shell
$ fwupdmgr get-devices --show-all-devices
System Product Name
│
├─AMD Ryzen 9 3900X 12-Core Processor            :
│     Device ID:          4bde70ba4e39b28f9eab1628f9dd6e6244c03027
│     Vendor:             AMD
│     GUIDs:              b9a2dd81-159e-5537-a7db-e7101d164d3f ← cpu
│                         997104da-ec11-5e04-9d94-ac5501fbf609 ← CPUID\PRO_0&FAM_17
│                         b54f3c89-8a41-5b9f-9a5c-0367c19a9451 ← CPUID\PRO_0&FAM_17&MOD_71
│                         d951a23c-a920-50ac-ad89-0660aa0be95c ← CPUID\PRO_0&FAM_17&MOD_71&STP_0
│     Device Flags:       • Internal device
│   
├─Force MP600:
│     Device ID:          71b677ca0f1bc2c5b804fa1d59e52064ce589293
│     Summary:            NVM Express Solid State Drive
│     Current version:    EGFM11.3
│     Vendor:             Phison Electronics Corporation (NVME:0x1987)
│     Serial Number:      20068229000128554D0C
│     GUIDs:              95711de3-8cc3-5db4-b813-6f4535e75c66 ← NVME\VEN_1987&DEV_5016&SUBSYS_19875016&REV_01
│                         12c32247-6129-573a-8e42-f8bf95e33f43 ← NVME\VEN_1987&DEV_5016&SUBSYS_19875016
│                         e0beed4b-7f9d-58e2-ae74-0dd4cf071aad ← NVME\VEN_1987&DEV_5016&REV_01
│                         3cb8c6f1-e686-5c7b-9ddf-cd66cd34350f ← NVME\VEN_1987&DEV_5016
│                         346b0c59-8802-541f-9270-2952ae216f2b ← Force MP600
│     Device Flags:       • Internal device
│                         • Updatable
│                         • System requires external power source
│                         • Needs shutdown after installation
│                         • Device is usable for the duration of the update
│   
├─Navi 10 [Radeon RX 5600 OEM/5600 XT / 5700/5700 XT]:
│     Device ID:          59fdb4840792b490058110f9228c05a14e30bd96
│     Current version:    c0
│     Vendor:             Advanced Micro Devices, Inc. [AMD/ATI] (PCI:0x1002)
│     GUIDs:              f276f287-498e-57d5-9da9-936f771b3147 ← PCI\VEN_1002&DEV_731F&SUBSYS_10020B36&REV_C0
│                         fe4eb599-a545-55f6-aed3-2879ab961bb1 ← PCI\VEN_1002&DEV_731F&SUBSYS_10020B36
│                         a60ba908-7e8c-5e15-8b61-ae67554e8e33 ← PCI\VEN_1002&DEV_731F&REV_C0
│                         dd0fc8c4-f037-5a78-bdea-c85eb0fe68fa ← PCI\VEN_1002&DEV_731F
│     Device Flags:       • Internal device
│                         • Cryptographic hash verification is available
│   
├─Samsung SSD 960 PRO 512GB:
│     Device ID:          310f45f1f223064b5c16bf6dff31146755a64480
│     Summary:            NVM Express Solid State Drive
│     Current version:    2B6QCXP7
│     Vendor:             Samsung Electronics Co Ltd (NVME:0x144D)
│     Serial Number:      S3EWNX0J515531N
│     GUIDs:              030c853f-259a-57a0-b3fb-1c66100db94b ← NVME\VEN_144D&DEV_A804&SUBSYS_144DA801&REV_00
│                         aed4d3c0-fd97-5e46-a32f-ff35e0692f6d ← NVME\VEN_144D&DEV_A804&SUBSYS_144DA801
│                         f87b9ac8-1cb3-5c0a-ae57-7144f211fe5e ← NVME\VEN_144D&DEV_A804&REV_00
│                         5b3df2da-f745-5fd0-81de-5dafd7f0bf8c ← NVME\VEN_144D&DEV_A804
│                         8c34a061-a08c-5a4c-81f4-813691222595 ← Samsung SSD 960 PRO 512GB
│     Device Flags:       • Updatable
│                         • System requires external power source
│                         • Needs a reboot after installation
│   
├─Samsung SSD 970 EVO Plus 1TB:
│     Device ID:          04e17fcf7d3de91da49a163ffe4907855c3648be
│     Summary:            NVM Express Solid State Drive
│     Current version:    2B2QEXM7
│     Vendor:             Samsung Electronics Co Ltd (NVME:0x144D)
│     Serial Number:      S4EWNF0MA01556K
│     GUIDs:              0b4d773a-7ac3-58c1-a541-e22ef1cdfe02 ← NVME\VEN_144D&DEV_A808&SUBSYS_144DA801&REV_00
│                         c9d531ea-ee7d-5562-8def-c64d0d144813 ← NVME\VEN_144D&DEV_A808&SUBSYS_144DA801
│                         6e54c992-d302-59ab-b454-2d26ddd63e6d ← NVME\VEN_144D&DEV_A808&REV_00
│                         47335265-a509-51f7-841e-1c94911af66b ← NVME\VEN_144D&DEV_A808
│                         f05ecac9-5cc2-5f77-b23b-a08ad8879ada ← Samsung SSD 970 EVO Plus 1TB
│     Device Flags:       • Updatable
│                         • System requires external power source
│                         • Needs a reboot after installation
│   
├─SanDisk SDSSDXPS480G:
│     Device ID:          74861bcb9108e6a21cde76f31dad912828a79c03
│     Summary:            ATA Drive
│     Current version:    X21200RL
│     Vendor:             SanDisk (ATA:0x15B7)
│     Serial Number:      153494402806
│     GUIDs:              92a0acec-17be-583f-bb5b-5b2253993871 ← IDE\SanDisk_SDSSDXPS480G____________________X21200RL
│                         2d967406-4294-5b65-8797-aea317ee6af5 ← IDE\0SanDisk_SDSSDXPS480G____________________
│                         a0b732d9-d6fc-506c-8749-13108ff87d21 ← SanDisk SDSSDXPS480G
│     Device Flags:       • Updatable
│                         • System requires external power source
│                         • Needs a reboot after installation
│   
├─System Firmware:
│ │   Device ID:          32302d7aed57230d1cb752b79eb1d0c908c49040
│ │   Current version:    9735
│ │   Minimum Version:    9735
│ │   Vendor:             System manufacturer (DMI:American Megatrends Inc.)
│ │   GUIDs:              d9aa4a14-e8aa-5abd-9687-3f7ee6de78ef
│ │                       230c8b18-8d9b-53ec-838b-6cfc0383493a ← main-system-firmware
│ │                       bc265207-7e11-50cd-9e02-61fac49f4882 ← UEFI\RES_{D9AA4A14-E8AA-5ABD-9687-3F7EE6DE78EF}
│ │   Device Flags:       • Internal device
│ │                       • Updatable
│ │                       • System requires external power source
│ │                       • Needs a reboot after installation
│ │                       • Cryptographic hash verification is available
│ │                       • Device is usable for the duration of the update
│ │ 
│ └─UEFI dbx:
│       Device ID:        362301da643102b9f38477387e2193e57abaa590
│       Summary:          UEFI Revocation Database
│       Current version:  77
│       Minimum Version:  77
│       Vendor:           UEFI:Linux Foundation
│       Install Duration: 1 second
│       GUIDs:            fda6234b-adcb-5105-8515-9af647d29775 ← UEFI\CRT_D7F66BE77CEF858C174BF4338A99263C8795B74E02026411F5F532F716AE3263
│                         f8ff0d50-c757-5dc3-951a-39d86e16f419 ← UEFI\CRT_D7F66BE77CEF858C174BF4338A99263C8795B74E02026411F5F532F716AE3263&ARCH_X64
│                         c6682ade-b5ec-57c4-b687-676351208742 ← UEFI\CRT_A1117F516A32CEFCBA3F2D1ACE10A87972FD6BBE8FE0D0B996E09E65D802A503
│                         f8ba2887-9411-5c36-9cee-88995bb39731 ← UEFI\CRT_A1117F516A32CEFCBA3F2D1ACE10A87972FD6BBE8FE0D0B996E09E65D802A503&ARCH_X64
│                         7d5759e5-9aa0-5f0c-abd6-7439bb11b9f6 ← UEFI\CRT_ED1FE72CB9CA31C9AF5B757AFCD733323D675825032E6CED7FE1AE9EB767998C
│                         0c7691e1-b6f2-5d71-bc9c-aabee364c916 ← UEFI\CRT_ED1FE72CB9CA31C9AF5B757AFCD733323D675825032E6CED7FE1AE9EB767998C&ARCH_X64
│       Device Flags:     • Internal device
│                         • Updatable
│                         • Needs a reboot after installation
│     
├─TOSHIBA HDWD130:
│     Device ID:          83298a50b68b28fbd7fe4219976f20d4c9e9e057
│     Summary:            ATA Drive
│     Current version:    MX6OACF0
│     Vendor:             Toshiba (ATA:0x1179)
│     Serial Number:      Z7440XBAS
│     GUIDs:              5d66898d-8ef4-5172-a7b7-f32d0fb9e988 ← IDE\TOSHIBA_HDWD130_________________________MX6OACF0
│                         d7c062f9-696d-5f9f-a9ac-7b6f8513250b ← IDE\0TOSHIBA_HDWD130_________________________
│                         9baff420-a2c7-57dc-9c33-c58103464100 ← TOSHIBA HDWD130
│     Device Flags:       • Updatable
│                         • System requires external power source
│                         • Needs a reboot after installation
│   
├─TPM:
│ │   Device ID:          c6a80ac3a22083423992a3cb15018989f37834d6
│ │   Summary:            TPM 2.0 Device
│ │   Current version:    3.37.0.5
│ │   Vendor:             AMD (TPM:AMD)
│ │   GUIDs:              ff71992e-52f7-5eea-94ef-883e56e034c6 ← system-tpm
│ │                       9305de1c-1e12-5665-81c4-37f8e51219b8 ← TPM\VEN_AMD&DEV_0001
│ │                       78a291ae-b499-5b0f-8f1d-74e1fefd0b1c ← TPM\VEN_AMD&MOD_AMD
│ │                       65a3fced-b423-563f-8098-bf5c329fc063 ← TPM\VEN_AMD&DEV_0001&VER_2.0
│ │                       5e704f0d-83cb-5364-8384-f46d725a23b8 ← TPM\VEN_AMD&MOD_AMD&VER_2.0
│ │   Device Flags:       • Internal device
│ │ 
│ └─Event Log:
│       Device ID:        58bd405f31c48e6eca290b425f530a94c91e955c
│       GUID:             a25657fe-b5dc-5be0-8b78-8b9dfec678ff ← system-tpm-eventlog
│       Device Flags:     • Internal device
│     
└─WDC WD7501AALS-00J7B0:
      Device ID:          1cb4548db4ea4d3a939a960a602d905d8f035e7a
      Summary:            ATA Drive
      Current version:    05.00K05
      Vendor:             Western Digital (ATA:0x101C)
      Serial Number:      WD-WMATV0156388
      GUIDs:              34d34cc5-5ad6-5ec9-9adb-8e8c253d0a61 ← IDE\WDC_WD7501AALS-00J7B0___________________05.00K05
                          29290fa4-061c-577a-94d0-4ff7bf340567 ← IDE\0WDC_WD7501AALS-00J7B0___________________
                          33d7d102-aa9b-5013-9d29-bb028a8f2d1e ← WDC WD7501AALS-00J7B0
      Device Flags:       • Updatable
                          • System requires external power source
                          • Needs a reboot after installatio
```

**Additional questions**
- Operating system and version: Fedora 33
- Have you tried rebooting? yes
- Is this a regression? no

**Comment by edwintorok** [comment]
https://github.com/fwupd/fwupd/wiki/Host-security-ID-runtime-issues should also be updated to point to the zram-as-swap as an alternative to encrypting swap.
(FWIW I've also converted a Ubuntu workstation to zram-on-swap and it has visibly improved system responsiveness  when using Firefox on sites that eat up massive amounts of RAM - such as leaving a Jenkins tab open. It is better than turning off swap, which would cause pagecache to be dropped when under memory pressure, which would lead to page-cache thrashing.
Unless you want to hibernate, or have very limited amounts of RAM I don't think disk based swap is needed/useful).

**Comment by hughsie** [comment]
I think https://github.com/fwupd/fwupd/pull/2529 addresses the zram case, no?

**Comment by edwintorok** [comment]
Thanks, it looks like the right approach (I haven't tried compiling it yet, but looking for `/dev/zram` as prefix should catch this, not sure how it'll be reported, reporting zram as encrypted isn't quite accurate, might be nicer to be a tri-state: encrypted, not encrypted, and not-applicable/internal/memory-only for zram?)

**Comment by edwintorok** [comment]
fwupd-1.5.1-1.fc33.x86_64 now shows this, with only HSI-3 and -U and -A being red:
```
Host Security ID: HSI:2 (v1.5.1)

HSI-1
✔ TPM v2.0:                      Found

HSI-2
✔ IOMMU:                         Enabled
✔ TPM PCR0 reconstruction:       Valid

HSI-3
✘ Suspend-to-idle:               Disabled
✘ Suspend-to-ram:                Enabled

Runtime Suffix -U
✘ Firmware updates:              Not supported

Runtime Suffix -A
✘ Firmware attestation:          Not supported

Runtime Suffix -!
✔ Linux kernel:                  Untainted
✔ Linux kernel lockdown:         Enabled
✔ Linux swap:                    Encrypted
✔ fwupd plugins:                 Untainted
```

**Comment by dnoliver** [comment]
I think I am running into this reported issue. 

I am using the same swap encryption configuration since F31, and before the firmware update manager reported my swap as encrypted. I mostly took the instructions from here https://wiki.archlinux.org/index.php/Dm-crypt/Swap_encryption

Versions are (this is Fedora IoT Edition):

```shell
[root@fitlet-1 ~]# cat /etc/redhat-release
Fedora release 33 (Thirty Three)

[root@fitlet-1 ~]# rpm -qa fwupd
fwupd-1.5.5-1.fc33.x86_64
```

Security report say that swap is not encrypted.

```shell
[root@fitlet-1 ~]# fwupdmgr security --force
Unknown                  [***************************************]
Host Security ID: HSI:2! (v1.5.5)

HSI-1
? Intel DCI debugger:            Disabled
? SPI BIOS region:               Locked
? SPI lock:                      Enabled
? SPI write:                     Disabled
? TPM v2.0:                      Found
? TXE manufacturing mode:        Locked
? TXE override:                  Locked
? TXE v0:3.1.70.2334:            Valid
? UEFI platform key:             Valid

HSI-2
? IOMMU:                         Enabled
? Intel DCI debugger:            Locked
? TPM PCR0 reconstruction:       Valid

HSI-3
? Suspend-to-ram:                Disabled
? Intel CET Enabled:             Not supported
? Pre-boot DMA protection:       Disabled
? Suspend-to-idle:               Disabled

HSI-4
? Intel SMAP:                    Enabled
? Encrypted RAM:                 Not supported

Runtime Suffix -!
? Linux kernel:                  Untainted
? Linux kernel lockdown:         Enabled
? fwupd plugins:                 Untainted
? Linux swap:                    Unencrypted

This system has HSI runtime issues.
 ? https://github.com/fwupd/fwupd/wiki/Host-security-ID-runtime-issues
```

My swap state is the following:

```shell
[root@fitlet-1 ~]# swapon -s
Filename                                Type            Size    Used    Priority
/dev/zram0                              partition       4194300 0       100
/dev/dm-1                               partition       12464124        0       -2

[root@fitlet-1 ~]# lsblk
NAME                     MAJ:MIN RM   SIZE RO TYPE  MOUNTPOINT
sda                        8:0    0 238.5G  0 disk
|-sda1                     8:1    0   200M  0 part  /boot/efi
|-sda2                     8:2    0   512M  0 part  /boot
`-sda3                     8:3    0 237.8G  0 part
  |-system-root          253:0    0  47.6G  0 lvm   /sysroot
  |-system-swap          253:1    0  11.9G  0 lvm   [SWAP]
  |-system-home          253:2    0  11.9G  0 lvm
  | `-luks-system-home   253:5    0  11.9G  0 crypt /var/home
  |-system-opt           253:3    0  47.6G  0 lvm
  | `-luks-system-opt    253:7    0  47.5G  0 crypt /var/opt
  `-system-docker        253:4    0 118.9G  0 lvm
    `-luks-system-docker 253:6    0 118.9G  0 crypt /var/lib/docker
zram0                    252:0    0     4G  0 disk  [SWAP]
```

So, with my configuration, I end up with 2 swap devices, which may be a problem.

But I guess the zram should pass this test based on the report from @edwintorok, an my system-swap should pass this test because it is encrypted. 

Should I file a bug for this?
Is my swap configuration wrong, and I should rather deactivate zram-swap and leave system-swap, or the other way around?

**Comment by hughsie** [comment]
> Is my swap configuration wrong

You only get the checkmark for encrypted swap if they're **all** encrypted.

**Comment by dnoliver** [comment]
I tested a little bit more, by disabling and enabling swap-on-zram and swap-on-disk, and running the security test after it.

**swap-on-zram** is identified as "encrypted" (not actually encrypted, but still means we do not dump secrets to disk)
**swap-on-disk** is identified as "not encrypted". This is weird, because I use the same swap encryption configuration since always, I did not found any new instructions saying that my config is no longer valid, `swapon` say I have a random `/dev/dm-X` device, but `lsblk` says my system-swap is not "crypt". Is it possible that the new detection logic based on `udisk2` thinks my swap is plain when it is not? also, it is possible that the old detection logic was identifying my swap as encrypted, but it was actually plain? 

Anyways, I think **swap-in-zram** is the way to go for me, and the default in Fedora IoT, so there is less configuration to do.
